src/main/java/org/truffleruby/core/rope/RopeOperations.java

```diff
@@ -657,4 +657,32 @@ public static Rope ropeFromBuilder(RopeBuilder byteList, CodeRange codeRange) {
         return create(byteList.getBytes(), byteList.getEncoding(), codeRange);
     }
 
+    public static boolean isAsciiOnly(byte[] bytes, Encoding encoding) {
+        if (!encoding.isAsciiCompatible()) {
+            return false;
+        }
+
+        for (int i = 0; i < bytes.length; i++) {
+            if (bytes[i] < 0) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public static boolean isInvalid(byte[] bytes, Encoding encoding) {
+        if (encoding.isAsciiCompatible()) {
+            final long packedLengthAndCodeRange = StringSupport.strLengthWithCodeRangeAsciiCompatible(encoding, bytes, 0 , bytes.length);
+            final CodeRange codeRange = CodeRange.fromInt(StringSupport.unpackArg(packedLengthAndCodeRange));
+
+            return codeRange == CR_BROKEN;
+        }
+
+        final long packedLengthAndCodeRange = StringSupport.strLengthWithCodeRangeNonAsciiCompatible(encoding, bytes, 0 , bytes.length);
+        final CodeRange codeRange = CodeRange.fromInt(StringSupport.unpackArg(packedLengthAndCodeRange));
+
+        return codeRange == CR_BROKEN;
```

Comment:
This looks like both branches are the same, should there be a
`strLengthWithCodeRange` that choses the right one based on the passed encoding?

Reply:
I already have `calculateCodeRangeAndLength`. I'll switch to that.